### PR TITLE
fix(470): autobump version

### PR DESCRIPTION
**Pull request by d2lam**
Auto bump version:
For example, if latest version is 1.7.3

If template.yaml says 1, it will create 1.7.4
If template.yaml says 1.7, it will create 1.7.4
If template.yaml says 1.7.3, it will create 1.7.4 (does not matter what the patch is)
If template.yaml says 1.8, it will create 1.8.0
If template.yaml says 2, it will create 2.0.0

Blocked by https://github.com/screwdriver-cd/models/pull/151
https://github.com/screwdriver-cd/models/pull/152
Related: https://github.com/screwdriver-cd/screwdriver/issues/470

### CODE REVIEW

**Comment by FenrirUnbound** [comment]
> We should only ship new versions if the template.yaml changes... Templates are just static content, there's no reason to rebuild

Ok, that makes a lot of sense. I was thinking along the lines of a shared step like `npm_install@1`. Assuming that it will work similar to how we intend templates to function, I can totally see this format working.

> We should be using the same learnings we got from Node.js and users should never have to bump their own patch version if they are changing something. That would make PRs impossible to merge one after another as each one would create conflicts with the next one.

Making the patch version automatic solves a lot of other edge cases, including multiple PRs to amend a template file. 

A semver-related question then: does this mean that any change to a template would either be a major change or a patch change? Phrased another way: what does a minor version bump look like?

**Comment by d2lam** [comment]
@stjohnjohnson  @FenrirUnbound  We implemented auto bump. 
For example, if latest version is `1.7.3`
If `template.yaml` says `1`, it will create `1.7.4`
If `template.yaml` says `1.7`, it will create `1.7.4`
If `template.yaml` says `1.8`, it will create `1.8.0`
If `template.yaml` says `2`, it will create `2.0.0`

File: plugins/templates/create.js [hunk 1]
```diff
@@ -26,18 +29,22 @@ module.exports = () => ({
...
 
             return Promise.all([
                 pipelineFactory.get(pipelineId),
                 templateFactory.list({ name })
             ]).then(([pipeline, templates]) => {
+                const VERSION_REGEX = schema.config.regex.VERSION;
+                const match = VERSION_REGEX.exec(request.payload.version);
+                const major = match[MAJOR_MATCH];
+                const minor = match[MINOR_MATCH];
+                const version = minor ? `${major}${minor}.0` : `${major}.0.0`;
```
Comment (FenrirUnbound) [inline plugins/templates/create.js]: It looks like the Regex match is wrong. Why isn't it `${major}.${minor}.0` ?

File: plugins/templates/create.js [hunk 2]
```diff
@@ -51,17 +58,28 @@ module.exports = () => ({
...
 
-                // If template name exists and has good permission, check the exact version
-                return templateFactory.get({ name, version })
-                    .then((template) => {
-                        // If the exact version exists, throw 409
-                        if (template) {
-                            throw boom.conflict(`Template ${name}@${version} already exists`);
-                        }
+                // If template name exists and has good permission, get the latest version for major and minor
+                const searchVersion = minor ? `${major}${minor}` : major;
```
Comment (FenrirUnbound) [inline plugins/templates/create.js]: `version` could be used instead of this.

File: plugins/templates/create.js (hunk 1 above)
Comment (d2lam) [inline plugins/templates/create.js]: second group includes the `.`: https://github.com/screwdriver-cd/data-schema/blob/master/config/regex.js#L11

Comment (FenrirUnbound) [inline plugins/templates/create.js]: The regex should return you just the minor version

File: plugins/templates/create.js (hunk 2 above)
Comment (d2lam) [inline plugins/templates/create.js]: `searchVersion` is used to get the latest of major & minor, it shouldn't include the patch. 
For example,
payload.version = 1.2.0
version = 1.2.0
searchVersion = 1.2

Comment (FenrirUnbound) [inline plugins/templates/create.js]: Does it not default to a 0 for the patch?

**Comment by FenrirUnbound** [comment]
This is a lot of logic in a Controller. Why have it all handled here instead of in a Model?

File: plugins/templates/create.js (hunk 2 above)
Comment (d2lam) [inline plugins/templates/create.js]: with `1.2`, `getTemplate`returns `1.2.latestPatch`, 
with `1.2.0`, `getTemplate` returns the exact version `1.2.0`

File: plugins/templates/create.js (hunk 1 above)
Comment (d2lam) [inline plugins/templates/create.js]: I'm not too good with regex, can you suggest a regex that would give that?

**Comment by d2lam** [comment]
@FenrirUnbound  That's possible. You mean to just have a helper inside templateFactory to get the correct version to create?

File: plugins/templates/create.js (hunk 1 above)
Comment (FenrirUnbound) [inline plugins/templates/create.js]: It just looks like a bug on the Regex matcher. I would have expected it to match *only* the version number and not include the delimiter

**Comment by FenrirUnbound** [comment]
> That's possible. You mean to just have a helper inside templateFactory to get the correct version to create?

Yeah. Some of the checks make sense to make here, like the permissions lookup. The business logic should be placed in the Models. A lot of the version patching and compiling could leverage the `semver` module.

**Comment by d2lam** [comment]
Yeah that makes sense. I will move some logic to models

**Comment by FenrirUnbound** [comment]
>  I will move some logic to models

👍  A lot of my concerns will vanish when the `semver` module is utilized. All of the version number handling goes away, and the logic will focus solely on the edge cases you're trying to capture

File: test/plugins/templates.test.js [hunk 3]
```diff
@@ -245,15 +237,6 @@ describe('template plugin test', () => {
             });
         });
 
-        it('returns 409 it is an existing version', () => {
```
Comment (d2lam) [inline test/plugins/templates.test.js]: Nope, since `create` will always bump version now.
